test/optim/test_optim.py

```diff
@@ -60,67 +60,93 @@ def _test_rosenbrock_sparse(
         scheduler_constructors=None,
         sparse_only=False,
         maximize=False,
+        multi_tensor=False
```

Comment:
We add a multi_tensor mode to the sparse test which adds a tensor of different datatype to ensure multiple optimizer param groups exist.

Reply:
@janeyx99 should we enable the multitensor test case for sparse sgd too?

Reply:
Yes, even more so, since we are supposed to support SGD for CUDA and CPU :') This should go in a different PR though.

https://github.com/pytorch/pytorch/issues/9674

This bug was long uncaught because foreach is not defaulted to on CPU (where it is not faster).  For SGD though, that is not the case.

Reply:
Sure, we could land this PR first then do SGD case in subsequent PR

Reply:
Actually @janeyx99 , can we just do it in this PR? The tests pass. That's because SGD `foreach` with sparse merely calls `add_` in a for loop. https://github.com/pytorch/pytorch/blob/98c8550158a4a79c4d39533a5331c5953f6ea279/torch/optim/sgd.py#L326-L327

Reply:
let’s keep it in a separate PR to maintain modularity. you can use ghstack to stack that as a follow up

Reply:
ok, done.